cmd/util/ledger/migrations/add_key_migration.go

```diff
@@ -0,0 +1,124 @@
+package migrations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/crypto/hash"
+	"github.com/onflow/flow-go-sdk/crypto"
+	"github.com/onflow/flow-go/cmd/util/ledger/util/registers"
+	"github.com/onflow/flow-go/fvm"
+	"github.com/onflow/flow-go/fvm/systemcontracts"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// AddKeyMigration adds a new key to the core contracts accounts
+type AddKeyMigration struct {
+	log zerolog.Logger
+
+	accountsToAddKeyTo map[common.Address]AddKeyMigrationAccountPublicKeyData
+	chainID            flow.ChainID
+}
+
+var _ AccountBasedMigration = &AddKeyMigration{}
+
+func NewAddKeyMigration(
+	chainID flow.ChainID,
+	key crypto.PublicKey,
+) *AddKeyMigration {
+
+	addresses := make(map[common.Address]AddKeyMigrationAccountPublicKeyData)
+	sc := systemcontracts.SystemContractsForChain(chainID).All()
+
+	for _, sc := range sc {
+		addresses[common.Address(sc.Address)] = AddKeyMigrationAccountPublicKeyData{
+			PublicKey: key,
+			HashAlgo:  hash.SHA3_256,
+		}
+	}
+
+	return &AddKeyMigration{
+		accountsToAddKeyTo: addresses,
+		chainID:            chainID,
+	}
+}
+
+type AddKeyMigrationAccountPublicKeyData struct {
+	PublicKey crypto.PublicKey
+	HashAlgo  hash.HashingAlgorithm
+}
+
+func (m *AddKeyMigration) InitMigration(
+	log zerolog.Logger,
+	_ *registers.ByAccount,
+	_ int,
+) error {
+	m.log = log.With().Str("component", "AddKeyMigration").Logger()
+	return nil
+}
+
+func (m *AddKeyMigration) Close() error {
+	return nil
+}
+
+func (m *AddKeyMigration) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	accountRegisters *registers.AccountRegisters,
+) error {
+
+	keyData, ok := m.accountsToAddKeyTo[address]
+	if !ok {
+		return nil
+	}
+
+	// Create all the runtime components we need for the migration
+	migrationRuntime, err := NewInterpreterMigrationRuntime(
+		accountRegisters,
+		m.chainID,
+		InterpreterMigrationRuntimeConfig{},
+	)
+
+	if err != nil {
+		return err
+	}
+
+	key := flow.AccountPublicKey{
+		PublicKey: keyData.PublicKey,
+		SignAlgo:  keyData.PublicKey.Algorithm(),
+		HashAlgo:  keyData.HashAlgo,
+		Weight:    fvm.AccountKeyWeightThreshold,
+	}
+
+	err = migrationRuntime.Accounts.AppendPublicKey(flow.ConvertAddress(address), key)
```

Comment:
Not sure if it is necessary but mentioning just in case.  Before appending new key to accounts, do we want to check if the key are already stored?

Reply:
For all core contract accounts they should already be there, but doesn't hurt to add a check.